server/infinispan/src/main/java/org/hawkular/apm/server/infinispan/InfinispanPropertyCache.java

```diff
@@ -0,0 +1,230 @@
+
+    private static final Logger log = Logger.getLogger(InfinispanPropertyCache.class.getName());
+
+    protected static final String CACHE_NAME = "property";
+
+    @Resource(lookup = "java:jboss/infinispan/APM")
+    private CacheContainer cacheContainer;
+
+    private Cache<String, Property<?>> propertyCache;
+    private CacheEventListener eventListener;
+
+
+    public InfinispanPropertyCache() {
+        this.eventListener = new CacheEventListener();
+    }
+
+    public InfinispanPropertyCache(Cache<String, Property<?>> cache) {
+        this();
+        this.propertyCache = cache;
+        this.propertyCache.addListener(eventListener);
+    }
+
+    @Override
+    @PostConstruct
+    public void init() {
+        if (propertyCache != null) {
+            return;
+        }
+
+        // If cache container not already provisions, then must be running outside of a JEE
+        // environment, so create a default cache container
+        if (cacheContainer == null) {
+            log.fine("Using default cache");
+            propertyCache = InfinispanCacheManager.getDefaultCache(CACHE_NAME);
+        } else {
+            log.fine("Using container provided cache");
+            propertyCache = cacheContainer.getCache(CACHE_NAME);
+        }
+
+        propertyCache.addListener(eventListener);
+    }
+
+    @Override
+    public Property get(String tenantId, String id) {
+        Property<?> property = propertyCache.get(id);
+
+        if (property == null) {
+            property = getFromSystemOrEnvVariables(id);
+            if (property == null) {
+                return null;
+            }
+
+            try {
+                store(tenantId, property);
+            } catch (CacheException e) {
+                log.severe(String.format("Could not store environmental property: %s to cache", property));
+            }
+        }
+
+        if (log.isLoggable(Level.FINEST)) {
+            log.finest("Get property [id=" + id + "] = " + property);
+        }
+
+        return property;
+    }
+
+    @Override
+    public void store(String tenantId, List<Property<?>> properties) throws CacheException {
+        if (cacheContainer != null) {
+            propertyCache.startBatch();
+        }
+
+        for (Property<?> property: properties) {
+            store(tenantId, property);
+        }
+
+        if (cacheContainer != null) {
+            propertyCache.endBatch(true);
+        }
+    }
+
+    @Override
+    public void store(String tenantId, Property<?> property) throws CacheException {
+        if (log.isLoggable(Level.FINEST)) {
+            log.finest("Store property [id=" + property.getId() + "]: " + property);
+        }
+
+        propertyCache.put(property.getId(), property);
+    }
+
+    @Override
+    public void remove(String tenantId, String id) {
+        Property<?> remove = propertyCache.remove(id);
+
+        if (remove != null) {
+            eventListener.observersMap.remove(id);
```

Comment:
I think removal of the observers should be done by the client - as they may add the property back, expecting the observers to still be registered. i.e. if the client adds the observer, it should also be responsible for removing it.